src/isa/riscv64/system/mmu.c

```diff
@@ -114,10 +138,112 @@ static inline bool check_permission(PTE *pte, bool ok, vaddr_t vaddr, int type)
   }
   return true;
 }
+#ifdef CONFIG_RVH
+bool has_two_stage_translation(){
+  return hld_st || (mstatus->mprv && mstatus->mpv) || cpu.v;
+}
+
+void raise_G_ex(paddr_t gpaddr, vaddr_t vaddr, int type){
```

Comment:
The name G_ex is too simple to understand for others. just use guest_excep. More words is ok and better.